File: frontend/apps/android/app/src/main/java/app/pantopus/android/data/api/models/mailbox/v2/CommunityDetailDto.kt
@file:Suppress("PackageNaming", "LongParameterList", "MagicNumber")

package app.pantopus.android.data.api.models.mailbox.v2

import app.pantopus.android.data.api.models.common.JsonValue

/** HOA/neighborhood-group seal rendered in the badge card. */
data class CommunityGroupInfo(
    val name: String,
    val tagline: String?,
    val founded: String?,
    val role: String?,
    val membershipSince: String?,
    val memberCount: Int?,
    val isVerified: Boolean,
)

/** When/where/bring chunk per the A17.4 design's Event details card. */
data class CommunityEventInfo(
    val dayLabel: String?,
    val dateLabel: String?,
    val timeRange: String?,
    val location: String?,
    val locationNote: String?,
    val distanceLabel: String?,
    val bringItems: List<String>,
    val weatherSummary: String?,
    val weatherTemperatureF: Int?,
)

/** One attendee rendered in the strip. */
data class CommunityAttendee(
    val id: String,
    val displayName: String,
    val initials: String,
    val blockLabel: String?,
    val isVerified: Boolean,
)

/** Cross-link card pointing at the related Pulse thread. */
data class CommunityPulseThread(
    val threadId: String,
    val title: String,
    val replyCount: Int,
    val lastReplyAuthor: String?,
    val lastReplyPreview: String?,
    val lastReplyAge: String?,
)

/** Tri-state RSVP — mirrors the design's chip row. */
enum class CommunityRsvpStatus(val wire: String) {
    Going("going"),
    Maybe("maybe"),
    NotGoing("not_going"),
    Undecided("undecided"),
    ;

    companion object {
        fun fromWire(value: String?): CommunityRsvpStatus =
            when (value?.lowercase()) {
                "going", "will_attend" -> Going
                "maybe" -> Maybe
                "not_going", "declined" -> NotGoing
                else -> Undecided
            }
    }
}

/**
 * Community detail sub-payload decoded from `mail.object_payload` when
 * `mail_type == "community"`. Backend stores this as untyped JSON shaped
 * by the `CommunityMailItem` table (`backend/database/schema.sql:5350`).
 * The RSVP-going state flows back from
 * `POST /api/mailbox/v2/community/rsvp`
 * (`backend/routes/mailboxV2Phase3.js:746`).
 * [decodeFromObjectPayload] returns null when the payload doesn't carry
 * a community item id.
 */
data class CommunityDetailDto(
    val communityItemId: String,
    val group: CommunityGroupInfo,
    val event: CommunityEventInfo?,
    val attendees: List<CommunityAttendee>,
    val attendeeCount: Int,
    val attendeesFromBlock: Int?,
    val pulseThread: CommunityPulseThread?,
    val rsvp: CommunityRsvpStatus,
) {
    companion object {
        @Suppress("UNCHECKED_CAST", "LongMethod", "CyclomaticComplexMethod")
        fun decodeFromObjectPayload(payload: JsonValue?): CommunityDetailDto? {
            if (payload == null) return null
            val itemId =
                (payload["community_item_id"] as? String)
                    ?: (payload["id"] as? String) ?: return null
            if (itemId.isEmpty()) return null
            val groupRaw =
                (payload["group"] as? Map<String, Any?>)
                    ?: (payload["community"] as? Map<String, Any?>)
                    ?: emptyMap<String, Any?>()
            val group =
                CommunityGroupInfo(
                    name = (groupRaw["name"] as? String) ?: "Neighborhood group",
                    tagline = groupRaw["tagline"] as? String,
                    founded = groupRaw["founded"] as? String,
                    role =
                        (groupRaw["role"] as? String)
                            ?: (groupRaw["membership_role"] as? String),
                    membershipSince =
                        (groupRaw["membership_since"] as? String)
                            ?: (groupRaw["since"] as? String),
                    memberCount = (groupRaw["member_count"] as? Number)?.toInt(),
                    isVerified = (groupRaw["verified"] as? Boolean) ?: false,
                )
            val eventRaw = payload["event"] as? Map<String, Any?>
            val event =
                eventRaw?.let { evt ->
                    val weather = evt["weather"] as? Map<String, Any?>
                    val whenMap = evt["when"] as? Map<String, Any?>
                    val whereStr =
                        (evt["location"] as? String)
                            ?: (evt["where"] as? String)
                    val bring =
                        (evt["bring"] as? List<*>)
                            ?.mapNotNull { it as? String } ?: emptyList()
                    CommunityEventInfo(
                        dayLabel =
                            (evt["day_label"] as? String)
                                ?: (whenMap?.get("day") as? String),
                        dateLabel =
                            (evt["date_label"] as? String)
                                ?: (whenMap?.get("date") as? String),
                        timeRange =
                            (evt["time_range"] as? String)
                                ?: (whenMap?.get("range") as? String),
                        location = whereStr,
                        locationNote =
                            (evt["location_note"] as? String)
                                ?: (evt["where_note"] as? String),
                        distanceLabel = evt["distance_label"] as? String,
                        bringItems = bring,
                        weatherSummary = weather?.get("summary") as? String,
                        weatherTemperatureF =
                            (weather?.get("temperature_f") as? Number)?.toInt()
                                ?: (weather?.get("temp") as? Number)?.toInt(),
                    )
                }
            val attendeesRaw = payload["attendees"] as? List<*> ?: emptyList<Any?>()
            val attendees =
                attendeesRaw.mapNotNull { entry ->
                    val map = entry as? Map<String, Any?> ?: return@mapNotNull null
                    val name =
                        (map["display_name"] as? String)
                            ?: (map["name"] as? String) ?: return@mapNotNull null
                    if (name.isEmpty()) return@mapNotNull null
                    val initials = (map["initials"] as? String) ?: makeInitials(name)
                    CommunityAttendee(
                        id = (map["id"] as? String) ?: java.util.UUID.randomUUID().toString(),
                        displayName = name,
                        initials = initials,
                        blockLabel =
                            (map["block_label"] as? String)
                                ?: (map["block"] as? String),
                        isVerified = (map["verified"] as? Boolean) ?: true,
                    )
                }
            val attendeeCount =
                (payload["attendee_count"] as? Number)?.toInt()
                    ?: (payload["rsvp_count"] as? Number)?.toInt()
                    ?: attendees.size
            val attendeesFromBlock = (payload["attendees_from_block"] as? Number)?.toInt()
            val threadRaw = payload["pulse_thread"] as? Map<String, Any?>
            val pulseThread =
                threadRaw?.let { td ->
                    val threadId =
                        (td["thread_id"] as? String)
                            ?: (td["id"] as? String) ?: return@let null
                    val title = td["title"] as? String ?: return@let null
                    val last = td["last_reply"] as? Map<String, Any?>
                    CommunityPulseThread(
                        threadId = threadId,
                        title = title,
                        replyCount =
                            (td["reply_count"] as? Number)?.toInt()
                                ?: (td["count"] as? Number)?.toInt() ?: 0,
                        lastReplyAuthor =
                            (last?.get("author") as? String)
                                ?: (last?.get("who") as? String),
                        lastReplyPreview = last?.get("preview") as? String,
                        lastReplyAge =
                            (last?.get("age") as? String)
                                ?: (last?.get("when") as? String),
                    )
                }
            return CommunityDetailDto(
                communityItemId = itemId,
                group = group,
                event = event,
                attendees = attendees,
                attendeeCount = attendeeCount,
                attendeesFromBlock = attendeesFromBlock,
                pulseThread = pulseThread,
                rsvp = CommunityRsvpStatus.fromWire(payload["rsvp_status"] as? String),
            )
        }

        private fun makeInitials(name: String): String {
            val parts = name.split(" ").take(2)
            return parts.mapNotNull { it.firstOrNull()?.uppercaseChar()?.toString() }
                .joinToString("")
                .ifEmpty { "·" }
        }
    }
}

File: frontend/apps/ios/Pantopus/Core/Networking/Models/Mailbox/V2/CommunityDetailDTO.swift
//
//  CommunityDetailDTO.swift
//  Pantopus
//
//  T6.5d (P22) — Community (A17.4) sub-payload decoded from
//  `mail.object_payload` when `mail_type == "community"`. Backend
//  stores this as untyped JSON shaped by the `CommunityMailItem`
//  table (schema at `backend/database/schema.sql:5350`). The
//  RSVP-going state flows back from the existing
//  `POST /api/mailbox/v2/community/rsvp` route
//  (`backend/routes/mailboxV2Phase3.js:746`) — the decoder is
//  defensive and `decode(from:)` returns nil when the payload
//  doesn't carry a community item id.
//

import Foundation

/// HOA/neighborhood-group seal rendered in the badge card.
public struct CommunityGroupInfo: Sendable, Hashable {
    public let name: String
    public let tagline: String?
    public let founded: String?
    public let role: String?
    public let membershipSince: String?
    public let memberCount: Int?
    public let isVerified: Bool

    public init(
        name: String,
        tagline: String?,
        founded: String?,
        role: String?,
        membershipSince: String?,
        memberCount: Int?,
        isVerified: Bool
    ) {
        self.name = name
        self.tagline = tagline
        self.founded = founded
        self.role = role
        self.membershipSince = membershipSince
        self.memberCount = memberCount
        self.isVerified = isVerified
    }
}

/// When/where/bring chunk per the A17.4 design's Event details card.
public struct CommunityEventInfo: Sendable, Hashable {
    public let dayLabel: String?
    public let dateLabel: String?
    public let timeRange: String?
    public let location: String?
    public let locationNote: String?
    public let distanceLabel: String?
    public let bringItems: [String]
    public let weatherSummary: String?
    public let weatherTemperatureF: Int?

    public init(
        dayLabel: String?,
        dateLabel: String?,
        timeRange: String?,
        location: String?,
        locationNote: String?,
        distanceLabel: String?,
        bringItems: [String],
        weatherSummary: String?,
        weatherTemperatureF: Int?
    ) {
        self.dayLabel = dayLabel
        self.dateLabel = dateLabel
        self.timeRange = timeRange
        self.location = location
        self.locationNote = locationNote
        self.distanceLabel = distanceLabel
        self.bringItems = bringItems
        self.weatherSummary = weatherSummary
        self.weatherTemperatureF = weatherTemperatureF
    }
}

/// One attendee rendered in the strip. Backend wire shape is loose;
/// `displayName` is always present, the rest are best-effort.
public struct CommunityAttendee: Sendable, Hashable, Identifiable {
    public let id: String
    public let displayName: String
    public let initials: String
    public let blockLabel: String?
    public let isVerified: Bool

    public init(id: String, displayName: String, initials: String, blockLabel: String?, isVerified: Bool) {
        self.id = id
        self.displayName = displayName
        self.initials = initials
        self.blockLabel = blockLabel
        self.isVerified = isVerified
    }
}

/// Cross-link card pointing at the related Pulse thread.
public struct CommunityPulseThread: Sendable, Hashable {
    public let threadId: String
    public let title: String
    public let replyCount: Int
    public let lastReplyAuthor: String?
    public let lastReplyPreview: String?
    public let lastReplyAge: String?

    public init(
        threadId: String,
        title: String,
        replyCount: Int,
        lastReplyAuthor: String?,
        lastReplyPreview: String?,
        lastReplyAge: String?
    ) {
        self.threadId = threadId
        self.title = title
        self.replyCount = replyCount
        self.lastReplyAuthor = lastReplyAuthor
        self.lastReplyPreview = lastReplyPreview
        self.lastReplyAge = lastReplyAge
    }
}

/// Tri-state RSVP — mirrors the design's Going / Maybe / Can't make it
/// chip row. Wire string is lowercase per the backend convention.
public enum CommunityRsvpStatus: String, Sendable, Hashable {
    case going
    case maybe
    case notGoing = "not_going"
    case undecided

    public init(wire: String?) {
        switch wire?.lowercased() {
        case "going", "will_attend": self = .going
        case "maybe": self = .maybe
        case "not_going", "declined": self = .notGoing
        default: self = .undecided
        }
    }
}

/// Community detail payload — drives the A17.4 variant body.
public struct CommunityDetailDTO: Sendable, Hashable {
    /// PK of the `CommunityMailItem` row. Required so the RSVP mutation
    /// can find the right item.
    public let communityItemId: String
    public let group: CommunityGroupInfo
    public let event: CommunityEventInfo?
    public let attendees: [CommunityAttendee]
    public let attendeeCount: Int
    public let attendeesFromBlock: Int?
    public let pulseThread: CommunityPulseThread?
    public let rsvp: CommunityRsvpStatus

    public init(
        communityItemId: String,
        group: CommunityGroupInfo,
        event: CommunityEventInfo?,
        attendees: [CommunityAttendee],
        attendeeCount: Int,
        attendeesFromBlock: Int?,
        pulseThread: CommunityPulseThread?,
        rsvp: CommunityRsvpStatus
    ) {
        self.communityItemId = communityItemId
        self.group = group
        self.event = event
        self.attendees = attendees
        self.attendeeCount = attendeeCount
        self.attendeesFromBlock = attendeesFromBlock
        self.pulseThread = pulseThread
        self.rsvp = rsvp
    }

    public static func decode(from value: JSONValue?) -> CommunityDetailDTO? {
        guard let dict = value?.dictValue else { return nil }
        guard let itemId = dict["community_item_id"]?.stringValue
            ?? dict["id"]?.stringValue, !itemId.isEmpty else {
            return nil
        }
        let groupDict = dict["group"]?.dictValue ?? dict["community"]?.dictValue ?? [:]
        let group = CommunityGroupInfo(
            name: groupDict["name"]?.stringValue ?? "Neighborhood group",
            tagline: groupDict["tagline"]?.stringValue,
            founded: groupDict["founded"]?.stringValue,
            role: groupDict["role"]?.stringValue ?? groupDict["membership_role"]?.stringValue,
            membershipSince: groupDict["membership_since"]?.stringValue
                ?? groupDict["since"]?.stringValue,
            memberCount: groupDict["member_count"]?.numberValue.map { Int($0) },
            isVerified: groupDict["verified"]?.boolValue ?? false
        )
        let eventDict = dict["event"]?.dictValue
        let event = eventDict.map { evt -> CommunityEventInfo in
            let weatherDict = evt["weather"]?.dictValue
            let bring = (evt["bring"]?.arrayValue ?? []).compactMap(\.stringValue)
            return CommunityEventInfo(
                dayLabel: evt["day_label"]?.stringValue
                    ?? evt["when"]?.dictValue?["day"]?.stringValue,
                dateLabel: evt["date_label"]?.stringValue
                    ?? evt["when"]?.dictValue?["date"]?.stringValue,
                timeRange: evt["time_range"]?.stringValue
                    ?? evt["when"]?.dictValue?["range"]?.stringValue,
                location: evt["location"]?.stringValue ?? evt["where"]?.stringValue,
                locationNote: evt["location_note"]?.stringValue
                    ?? evt["where_note"]?.stringValue,
                distanceLabel: evt["distance_label"]?.stringValue,
                bringItems: bring,
                weatherSummary: weatherDict?["summary"]?.stringValue,
                weatherTemperatureF: weatherDict?["temperature_f"]?.numberValue.map { Int($0) }
                    ?? weatherDict?["temp"]?.numberValue.map { Int($0) }
            )
        }
        let attendees: [CommunityAttendee] = (dict["attendees"]?.arrayValue ?? [])
            .compactMap { attn -> CommunityAttendee? in
                guard let a = attn.dictValue,
                      let name = a["display_name"]?.stringValue ?? a["name"]?.stringValue,
                      !name.isEmpty else { return nil }
                let id = a["id"]?.stringValue ?? UUID().uuidString
                let initials = a["initials"]?.stringValue
                    ?? Self.makeInitials(from: name)
                return CommunityAttendee(
                    id: id,
                    displayName: name,
                    initials: initials,
                    blockLabel: a["block_label"]?.stringValue ?? a["block"]?.stringValue,
                    isVerified: a["verified"]?.boolValue ?? true
                )
            }
        let attendeeCount = dict["attendee_count"]?.numberValue.map { Int($0) }
            ?? dict["rsvp_count"]?.numberValue.map { Int($0) }
            ?? attendees.count
        let attendeesFromBlock = dict["attendees_from_block"]?.numberValue.map { Int($0) }
        let threadDict = dict["pulse_thread"]?.dictValue
        let pulseThread = threadDict.flatMap { td -> CommunityPulseThread? in
            guard let threadId = td["thread_id"]?.stringValue ?? td["id"]?.stringValue,
                  let title = td["title"]?.stringValue else { return nil }
            let lastReply = td["last_reply"]?.dictValue
            return CommunityPulseThread(
                threadId: threadId,
                title: title,
                replyCount: td["reply_count"]?.numberValue.map { Int($0) }
                    ?? td["count"]?.numberValue.map { Int($0) }
                    ?? 0,
                lastReplyAuthor: lastReply?["author"]?.stringValue
                    ?? lastReply?["who"]?.stringValue,
                lastReplyPreview: lastReply?["preview"]?.stringValue,
                lastReplyAge: lastReply?["age"]?.stringValue
                    ?? lastReply?["when"]?.stringValue
            )
        }
        return CommunityDetailDTO(
            communityItemId: itemId,
            group: group,
            event: event,
            attendees: attendees,
            attendeeCount: attendeeCount,
            attendeesFromBlock: attendeesFromBlock,
            pulseThread: pulseThread,
            rsvp: CommunityRsvpStatus(wire: dict["rsvp_status"]?.stringValue)
        )
    }

    private static func makeInitials(from name: String) -> String {
        let parts = name.split(separator: " ").prefix(2)
        let result = parts.compactMap { $0.first.map(String.init) }.joined().uppercased()
        return result.isEmpty ? "·" : result
    }
}
